std/math/fp2.asm

```diff
@@ -0,0 +1,159 @@
+use std::math::ff::inverse;
+use std::convert::fe;
+use std::convert::int;
+use std::convert::expr;
+use std::field::modulus;
+use std::prover::eval;
+
+// An extension field element, represented as a tuple (a0, a1) of expressions (which evaluate to field elements).
+// An element of F_{p^2} is a polynomial a0 + a1 * x (mod x^2 - 7).
+// The polynomial x^2 - 7 is irreducible in both the Goldilocks and BN254 field
+// (which follows from 7 not having a square root in both fields).
+enum Fp2Expr {
+    Fp2(expr, expr)
+}
+
+// An element of F_{p^2}, but represented as a tuple of *evaluated* field elements.
+enum Fp2Value {
+    Fp2(fe, fe)
+}
+
+let zero_expr = Fp2Expr::Fp2(0, 0);
+let one_expr = Fp2Expr::Fp2(1, 0);
+
+let zero_value = Fp2Value::Fp2(0, 0);
+let one_value = Fp2Value::Fp2(1, 0);
+
+// Extension field addition
+let add_ext: Fp2Expr, Fp2Expr -> Fp2Expr = |a, b| match (a, b) {
+    (Fp2Expr::Fp2(a0, a1), Fp2Expr::Fp2(b0, b1)) => Fp2Expr::Fp2(
+        a0 + b0,
+        a1 + b1
+    )
+};
+
+// Extension field subtraction
+let sub_ext: Fp2Expr, Fp2Expr -> Fp2Expr = |a, b| match (a, b) {
+    (Fp2Expr::Fp2(a0, a1), Fp2Expr::Fp2(b0, b1)) => Fp2Expr::Fp2(
+        a0 - b0,
+        a1 - b1
+    )
+};
+
+// Extension field multiplication
+let mul_ext: Fp2Expr, Fp2Expr -> Fp2Expr = |a, b| match (a, b) {
+    (Fp2Expr::Fp2(a0, a1), Fp2Expr::Fp2(b0, b1)) => Fp2Expr::Fp2(
+        // Multiplication modulo the polynomial x^2 - 7. We'll use the fact
+        // that x^2 == 7 (mod x^2 - 7), so:
+        // (a0 + a1 * x) * (b0 + b1 * x) = a0 * b0 + 7 * a1 * b1 + (a1 * b0 + a0 * b1) * x (mod x^2 - 7)
+        a0 * b0 + 7 * a1 * b1,
+        a1 * b0 + a0 * b1
+    )
+};
+
+// Converts and Fp2Expr into an Fp2Value
+let eval_ext: Fp2Expr -> Fp2Value = query |a| match a {
+    Fp2Expr::Fp2(a0, a1) => Fp2Value::Fp2(eval(a0), eval(a1))
+};
+
+// Converts and Fp2Value into an Fp2Expr
+let expr_ext: Fp2Value -> Fp2Expr = |a| match a {
+    Fp2Value::Fp2(a0, a1) => Fp2Expr::Fp2(expr(a0), expr(a1))
+};
+
+// Extension field equality
+let eq_ext: Fp2Value, Fp2Value -> bool = |a, b| match (a, b) {
+    (Fp2Value::Fp2(a0, a1), Fp2Value::Fp2(b0, b1)) => (a0 == b0) && (a1 == b1)
+};
+
+// Field inversion (defined on fe instead of int)
+let inv_field: fe -> fe = |x| fe(inverse(int(x), modulus()));
```

Comment:
I think it might be better to qualify `inverse` instead of importing it, since we do not (yet) have the `pub` feature, so the fp2 module would also export a function called `inverse`